pkg/webhook/cluster.go

```diff
@@ -23,16 +34,60 @@ func (r *ClusterWebhook) SetupWebhookWithManager(mgr ctrl.Manager) error {
 
 var _ webhook.CustomValidator = &ClusterWebhook{}
 
+// fetchInfrastructureObject fetches the Infrastructure object from the cluster.
+func (r *ClusterWebhook) fetchInfrastructureObject(ctx context.Context) (*configv1.Infrastructure, error) {
+	infrastructureObjectKey := client.ObjectKey{Name: "cluster", Namespace: "default"}
+
+	infrastructureObject := configv1.Infrastructure{}
+	err := r.client.Get(ctx, infrastructureObjectKey, &infrastructureObject)
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch Infrastructure object: %w", err)
+	}
+
+	return &infrastructureObject, nil
+}
+
+// In openshift-cluster-api allow only one Cluster object to be created. This Cluster manages the cluster we are running on.
+func (r *ClusterWebhook) validateClusterName(ctx context.Context, cluster *v1beta1.Cluster) error {
+	if cluster.Namespace != openshiftCAPINamespace {
+		return nil
+	}
+
+	infrastructureObject, err := r.fetchInfrastructureObject(ctx)
+	if err != nil {
+		return fmt.Errorf("cluster in %s namespace must be named <infrastructure_id>. Failed to obtain name from Infrastructure object for validation: %w", openshiftCAPINamespace, err)
+	}
+
+	infrastructureName := infrastructureObject.Status.InfrastructureName
+	if cluster.ObjectMeta.Name != infrastructureName {
+		return fmt.Errorf("cluster name must be %s in %s namespace", infrastructureName, openshiftCAPINamespace)
+	}
+
+	return nil
+}
+
 // ValidateCreate implements webhook.Validator so a webhook will be registered for the type
 func (r *ClusterWebhook) ValidateCreate(ctx context.Context, obj runtime.Object) error {
 	cluster, ok := obj.(*v1beta1.Cluster)
 	if !ok {
 		panic("expected to get an of object of type v1beta1.Cluster")
 	}
-	switch cluster.Spec.InfrastructureRef.Kind {
-	case "AWSCluster", "AzureCluster", "GCPCluster", "IBMPowerVSCluster":
-	default:
-		return fmt.Errorf("unsupported cluster infra provider kind: %s", cluster.Spec.InfrastructureRef.Kind)
+
+	errs := []error{}
+	infrastructureRefPath := field.NewPath("spec", "infrastructureRef")
+	if cluster.Spec.InfrastructureRef != nil {
```

Comment:
This change seems unrelated to the other changes, why do we allow a nil infrastructure ref? When would it ever be created without one? /should there be an else here in case it is empty?

Reply:
I was fixing a panic and did not want to make it a required field. All the field in the spec are optional. I am not sure what kind of validation we want in openshift-cluster-api namespace vs other namespaces, or if we plan to support running other workload clusters.

Reply:
Makes sense! Thanks